feral-lambda-runtime/src/main/scala/feral/lambda/runtime/LambdaRuntime.scala

```diff
@@ -14,56 +14,54 @@
  * limitations under the License.
  */
 
-// TODO implement a runtime here
-// it should retrieve incoming events and handle them with the handler
-// it will run on a background fiber, whose lifecycle is controlled by the resource
-
 package feral.lambda
 package runtime
 
-import cats.Applicative
 import cats.syntax.all._
-import cats.effect.kernel.{Resource, Sync}
+import cats.effect.kernel.Sync
 import io.circe.Json
 import org.http4s.Method.POST
 import org.http4s.client.Client
 import org.http4s.circe.jsonEncoderWithPrinter
-
 import scala.concurrent.duration.FiniteDuration
 import java.util.concurrent.TimeUnit
 import org.http4s.{EntityEncoder, Uri}
 import org.http4s.client.dsl.Http4sClientDsl
 import io.circe._
-
-import java.time.Instant //safe to use in native?
 import cats.effect.kernel.Async
+import cats.effect.std.Env
+import org.http4s.implicits.http4sLiteralsSyntax
+
+// TODO apply function error handling
+  // How to run global or static code from handler?
+// TODO Custom AWS header models
+// TODO CognitoIdentity/ClientContext JSON encoding
 
 object FeralLambdaRuntime {
 
-  val LAMBDA_VERSION_DATE = "2018-06-01"
+  final val ApiVersion = "2018-06-01"
 
-  def apply[F[_]](client: Client[F])(handler: (Json, Context[F]) => F[Json])(implicit F: Async[F]): Resource[F, Unit] = {
-    F.background {
-      val runtimeUrl = getRuntimeUrl(LambdaReservedEnvVars.AWS_LAMBDA_RUNTIME_API)
-      implicit val jsonEncoder: EntityEncoder[F, Json] = jsonEncoderWithPrinter[F](Printer.noSpaces.copy(dropNullValues = true))
-      val http4sClientDsl = new Http4sClientDsl[F] {}
-      import http4sClientDsl._
-      (for {
-        request <- client.get(runtimeUrl)(LambdaRequest.fromResponse) // unsure how to deal with bad response
-        context <- createContext(request)
-        result <- handler(request.body, context)
-        invocationUrl = getInvocationUrl(LambdaReservedEnvVars.AWS_LAMBDA_RUNTIME_API, request.id)
-        _ <- client.successful(POST(result, invocationUrl))
-      } yield ()).foreverM
-    }
-  }.as(()) // how to handle Outcome error and cancellation?
+  def apply[F[_]](client: Client[F])(handler: (Json, Context[F]) => F[Json])(implicit F: Async[F]): F[Unit] = {
+    implicit val lambdaEnv: LambdaRuntimeEnv[F] = LambdaRuntimeEnv(Env.make) // maybe better way
+    implicit val jsonEncoder: EntityEncoder[F, Json] = jsonEncoderWithPrinter[F](Printer.noSpaces.copy(dropNullValues = true))
+    val http4sClientDsl = new Http4sClientDsl[F] {}
+    import http4sClientDsl._
+    (for {
+      runtimeApi <- lambdaEnv.lambdaRuntimeApi
+      request <- client.get(getRuntimeUrl(runtimeApi))(LambdaRequest.fromResponse)
+      context <- createContext(request)
+      result <- handler(request.body, context)
+      invocationUrl = getInvocationUrl(LambdaRuntimeEnv.AWS_LAMBDA_RUNTIME_API, request.id)
+      _ <- client.successful(POST(result, invocationUrl))
```

Comment:
`successful` returns a `Boolean`, which is currently being ignored. I wonder if `expect[Unit]` will work, so it will raise an exception on failure.